packages/compass-import-export/src/export/gather-fields.ts

```diff
@@ -0,0 +1,169 @@
+import type { DataService } from 'mongodb-data-service';
+import mongodbSchema from 'mongodb-schema';
+import type { SchemaField } from 'mongodb-schema';
+import { isInternalFieldPath } from 'hadron-document';
+import type { Document } from 'mongodb';
+
+import { createDebug } from '../utils/logger';
+
+const debug = createDebug('export-json');
+
+type AnalyzeSchemaOptions = {
+  dataService: DataService;
+  ns: string;
+  abortSignal: AbortSignal;
+  filter?: Document;
+  sampleSize: number;
+};
+
+async function analyzeSchema({
+  dataService,
+  ns,
+  abortSignal,
+  filter,
+  sampleSize,
+}: AnalyzeSchemaOptions) {
+  try {
+    // TODO(COMPASS-6426): Should we use the other aspects of the query in this sample. (project/sort/limit/skip)
+    const docs = await dataService.sample(
+      ns,
+      {
+        query: filter,
+        ...(sampleSize
+          ? {
+              size: sampleSize,
+            }
+          : {}),
+      },
+      {
+        promoteValues: false,
+      },
+      {
+        abortSignal,
+      }
+    );
+    const schemaData = await mongodbSchema(docs);
+    schemaData.fields = schemaData.fields.filter(
+      ({ path }: { path: string }) => !isInternalFieldPath(path)
+    );
+
+    return schemaData;
+  } catch (err) {
+    if (dataService.isCancelError(err)) {
+      debug('caught background operation terminated error', err);
+      return null;
+    }
+
+    debug('schema analysis failed', err);
+    throw err;
+  }
+}
+
+type GatherFieldsOptions = {
+  dataService: DataService;
+  ns: string;
+  abortSignal: AbortSignal;
+  filter?: Document;
+  progressCallback: (index: number) => void;
```

Comment:
You can probably make this optional like we we now do with other progressCallbacks. If anything it makes testing easier when you really don't care about progress.